saleor/discount/utils.py

```diff
@@ -556,6 +729,41 @@ def get_variants_to_promotions_map(
     return rules_info_per_variant_and_promotion_id
 
 
+def fetch_promotion_rules_for_checkout(
+    checkout: Checkout,
+):
+    from ..graphql.discount.utils import PredicateObjectType, filter_qs_by_predicate
+
+    applicable_rules = []
+    promotions = Promotion.objects.active()
+    rules = (
+        PromotionRule.objects.filter(
+            Exists(promotions.filter(id=OuterRef("promotion_id")))
+        )
+        .exclude(order_predicate={})
+        .prefetch_related("channels")
+    )
+    rule_to_channel_ids_map = _get_rule_to_channel_ids_map(rules)
+
+    checkout_channel_id = checkout.channel_id
+    currency = checkout.channel.currency_code
+    checkout_qs = Checkout.objects.filter(pk=checkout.pk)
+    for rule in list(rules.iterator()):
```

Comment:
This means, that we will load ALL existing order rules to memory.

Reply:
I've made a test, on 5 mln of checkouts. With `https://github.com/pythonprofilers/memory_profiler`
```
In [5]: from memory_profiler import memory_usage
   ...: from time import sleep
   ...:
   ...: def f():
   ...:     # a function that with growing
   ...:     # memory consumption
   ...:     print(Checkout.objects.count())
   ...:     for a in list(Checkout.objects.iterator()):
   ...:         b = a.pk
   ...:     return None
   ```
   vs 
   ```
   In [5]: from memory_profiler import memory_usage
   ...: from time import sleep
   ...:
   ...: def f():
   ...:     # a function that with growing
   ...:     # memory consumption
   ...:     print(Checkout.objects.count())
   ...:     for a in Checkout.objects.iterator():
   ...:         b = a.pk
   ...:     return None
   ```
   
   `max(memory_usage(f))` returned `3737.390625` MiB for `list(iterator())`, and `235.671875` for `iterator()`
   
   Using `iterator` allows db to stream the data (ref:https://docs.djangoproject.com/en/5.0/ref/models/querysets/#with-server-side-cursors), but having `list(iterator())` will cause that we will load all data into worker memory.